### PR TITLE
feat: derive `IsmConfig` given an ism address

### PR DESCRIPTION
**Pull request by paulbalaji**
resolves https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3436

- `EvmIsmReader` class to derive `IsmConfig` given an EVM address
	- maps are enumerable here, unlike hooks, so no compromise required to iterate over domains
	- include addresses of all children in the returned `IsmConfig`
- updated `read-config.ts` script for reading both ism/hook configs before CLI integration
	- doesn't use multiprovider from envConfig, as it needs to know about non-EVM chains
- drive-by fix to `EvmHookReader` to respect `disableConcurrency` when looping through aggregation hooks

TODO:
- [x] tests
- [x] fix the tests

### CODE REVIEW

File: typescript/sdk/src/hook/read.ts [hunk 1]
```diff
@@ -132,9 +132,17 @@ export class EvmHookReader implements HookReader {
...
+    let hookConfigs = [];
+    if (this.disableConcurrency) {
+      for (const hookAddress of hooks) {
+        const config = await this.deriveHookConfig(hookAddress);
+        hookConfigs.push(config);
+      }
+    } else {
+      hookConfigs = await Promise.all(
+        hooks.map((hookAddress) => this.deriveHookConfig(hookAddress)),
+      );
```
[the remark this answers is not in the document]
Comment (yorhodes) [reply typescript/sdk/src/hook/read.ts:144]: https://github.com/celo-org/developer-tooling/blob/0c61e7e02c741fe10ecd1d733a33692d324cdc82/packages/sdk/base/src/async.ts#L128 dug this up from celo days

File: typescript/sdk/src/ism/read.ts [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+      domains[chainName] = await this.deriveIsmConfig(module);
+    };
+
+    if (this.disableConcurrency) {
+      for (const domainId of domainIds) {
+        await processDomainId(domainId);
+      }
+    } else {
+      await Promise.all(domainIds.map(processDomainId));
+    }
```
Comment (yorhodes) [inline typescript/sdk/src/ism/read.ts:119]: ditto here, pull into a util

File: typescript/sdk/src/ism/read.ts [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    if (this.disableConcurrency) {
+      for (const ismAddress of modules) {
+        const config = await this.deriveIsmConfig(ismAddress);
+        ismConfigs.push(config);
+      }
+    } else {
+      ismConfigs = await Promise.all(
+        modules.map((ismAddress) => this.deriveIsmConfig(ismAddress)),
+      );
+    }
```
Comment (yorhodes) [inline typescript/sdk/src/ism/read.ts:161]: ditto

**Review by yorhodes — APPROVED** [review]
nice work
any example output to share from your infra script?

**Comment by paulbalaji** [comment]
> nice work any example output to share from your infra script?

ya here's the example output of our top level ISM on ancient8
```json
{
  "address": "0x6E3387e12C6e181BF8e712eCa9c60ccEEaBD1c67",
  "type": "staticAggregationIsm",
  "modules": [
    {
      "address": "0x8f4BeB6552b76aA38Cd9994701c0Da7bC829648B",
      "owner": "0xa7ECcdb9Be08178f896c26b7BbD8C3D4E844d9Ba",
      "type": "pausableIsm",
      "paused": false
    },
    {
      "owner": "0xa7ECcdb9Be08178f896c26b7BbD8C3D4E844d9Ba",
      "address": "0xBD70Ea9D599a0FC8158B026797177773C3445730",
      "type": "domainRoutingIsm",
      "domains": {
        "avalanche": {
          "address": "0x979Dcd709dD63560894cB516174cA7909Dcf0D3D",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x131062a0a635bb73859F6616a7526135B6beb450",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x3fB8263859843bfFb02950c492D492cAe169f4Cf",
                "0x402E0f8c6E4210D408B6AC00D197D4A099fcd25A",
                "0x716A1d4D3166c6151B05cE0450e0D77d94588EaC"
              ],
              "threshold": 2
            },
            {
              "address": "0xc9fB2f3A26F5080cB5a0fE7B397694023e08c32e",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x3fB8263859843bfFb02950c492D492cAe169f4Cf",
                "0x402E0f8c6E4210D408B6AC00D197D4A099fcd25A",
                "0x716A1d4D3166c6151B05cE0450e0D77d94588EaC"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "arbitrum": {
          "address": "0x03E834bB1CdB247BD27EF7a004229F065aFdB784",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x485420cD1Be640CB3Bb91709addaCF1f30872Afe",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x092E1C19DA58e87EA65198785EE83867Fe4BB418",
                "0x4d966438fe9E2B1e7124c87bBB90cB4F0F6C59a1",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0xc2d68e109a7E80e12098d50ac4eF9fA7B3061684",
                "0xec68258A7c882AC2Fc46b81Ce80380054fFB4eF2"
              ],
              "threshold": 3
            },
            {
              "address": "0x79b1C9C10D8056fd076C986473521579576e25f6",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x092E1C19DA58e87EA65198785EE83867Fe4BB418",
                "0x4d966438fe9E2B1e7124c87bBB90cB4F0F6C59a1",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0xc2d68e109a7E80e12098d50ac4eF9fA7B3061684",
                "0xec68258A7c882AC2Fc46b81Ce80380054fFB4eF2"
              ],
              "threshold": 3
            }
          ],
          "threshold": 1
        },
        "viction": {
          "address": "0xb10200e048FafADb50AF2286C01cb1466F4fccC0",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x42DF3BD0dEaC280900463F2016D962fE33E16A07",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x1F87C368f8e05A85eF9126d984a980A20930cB9C",
                "0x4E53dA92cD5Bf0a032b6B4614b986926456756A7",
                "0xa3F93FE365bF99F431D8fdE740B140615e24f99B"
              ],
              "threshold": 2
            },
            {
              "address": "0xc056cC12752A206C02ffD02bF4ab903b8DF95ca6",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x1F87C368f8e05A85eF9126d984a980A20930cB9C",
                "0x4E53dA92cD5Bf0a032b6B4614b986926456756A7",
                "0xa3F93FE365bF99F431D8fdE740B140615e24f99B"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "mantapacific": {
          "address": "0x00cfc19eb97AbDcF3A724C771850b54a8c4616b4",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x424A84Dd6810Dd4430F4266e7F0F65160370D41e",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x14025fE092f5f8a401DD9819704D9072196D2125",
                "0x25b9A0961C51e74fd83295293Bc029131bF1E05A",
                "0x42B6De2edBaA62C2eA2309Ad85D20B3E37d38Acf",
                "0x521a3e6Bf8D24809fde1C1FD3494A859a16f132c",
                "0x8e668c97Ad76D0e28375275c41ece4972AB8a5bc",
                "0xa0eE95e280D46C14921e524B075d0C341e7ad1C8",
                "0xcc9a0b6dE7Fe314bd99223687d784730a75bB957"
              ],
              "threshold": 5
            },
            {
              "address": "0x59926F2e2C0fBd7a76b5CD101CfFF0b2A0A4e67F",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x14025fE092f5f8a401DD9819704D9072196D2125",
                "0x25b9A0961C51e74fd83295293Bc029131bF1E05A",
                "0x42B6De2edBaA62C2eA2309Ad85D20B3E37d38Acf",
                "0x521a3e6Bf8D24809fde1C1FD3494A859a16f132c",
                "0x8e668c97Ad76D0e28375275c41ece4972AB8a5bc",
                "0xa0eE95e280D46C14921e524B075d0C341e7ad1C8",
                "0xcc9a0b6dE7Fe314bd99223687d784730a75bB957"
              ],
              "threshold": 5
            }
          ],
          "threshold": 1
        },
        "optimism": {
          "address": "0xb527746eB58FE07735361226E2c6F00fdA8Fa6c9",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x8fc728014a142F099A70c81f95cFd55b2ab45441",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x20349eAdC6c72e94Ce38268b96692b1a5C20De4F",
                "0x22B1AD4322cDb5f2c76ebF4E5a93803d480fCf0D",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x5B7D47b76C69740462432f6a5a0ca5005E014157",
                "0x9636Fbe90B6816438327b0FBde435AA3c8eEDa15"
              ],
              "threshold": 2
            },
            {
              "address": "0xa9076E4bE74F541740D0958E7e49eD6CcdBc3a68",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x20349eAdC6c72e94Ce38268b96692b1a5C20De4F",
                "0x22B1AD4322cDb5f2c76ebF4E5a93803d480fCf0D",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x5B7D47b76C69740462432f6a5a0ca5005E014157",
                "0x9636Fbe90B6816438327b0FBde435AA3c8eEDa15"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "scroll": {
          "address": "0xE335B7e8E760b42C51145b5bbb1ed16D69197D3a",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x5b353218729d7477fED043929a14d15367eA9F40",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x276de8E2b88E659c4E5aD30D62d9De42c3DA3403",
                "0x37148DE77D9FA915e6F0A9B54bCdF5e6f53ca511",
                "0xaD557170a9f2f21C35E03de07CB30dcbcc3DfF63",
                "0xBAc4ac39F1D8b5eF15f26fDB1294a7c9aBa3f948"
              ],
              "threshold": 2
            },
            {
              "address": "0xC4518e79F27e1439b9BC49830d5cD3A5B0328254",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x276de8E2b88E659c4E5aD30D62d9De42c3DA3403",
                "0x37148DE77D9FA915e6F0A9B54bCdF5e6f53ca511",
                "0xaD557170a9f2f21C35E03de07CB30dcbcc3DfF63",
                "0xBAc4ac39F1D8b5eF15f26fDB1294a7c9aBa3f948"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "bsc": {
          "address": "0x79A7c7Fe443971CBc6baD623Fdf8019C379a7178",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x718901b7570c59241304758345aF297Ff47D83A1",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x570AF9B7B36568C8877eeBBA6c6727aA9DAB7268",
                "0x8292B1A53907eCE0f76Af8A50724e9492bcDc8A3",
                "0xEAF5cF9100f36A4BaEeA779F8745DDa86159103C"
              ],
              "threshold": 2
            },
            {
              "address": "0x8dac7Df615Cb7599aCfD4EC2CBd9c67Cd5DF6BF6",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x570AF9B7B36568C8877eeBBA6c6727aA9DAB7268",
                "0x8292B1A53907eCE0f76Af8A50724e9492bcDc8A3",
                "0xEAF5cF9100f36A4BaEeA779F8745DDa86159103C"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "inevm": {
          "address": "0x64DF9a02380bcd9d88b771771319818bff2d9664",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x2Fd7A5a277875C5d25237C4de796276B5852c0B8",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x6B1d09A97b813D53e9D4b7523DA36604C0B52242",
                "0x9aB11F38a609940153850Df611C9a2175dCfFe0f",
                "0xF9e35Ee88e4448a3673b4676A4E153e3584a08EB"
              ],
              "threshold": 2
            },
            {
              "address": "0x503b801999C335fAC2C306675DA8451799C52809",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x6B1d09A97b813D53e9D4b7523DA36604C0B52242",
                "0x9aB11F38a609940153850Df611C9a2175dCfFe0f",
                "0xF9e35Ee88e4448a3673b4676A4E153e3584a08EB"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "injective": {
          "address": "0x006a677cD9E24c9DCC906CA67fFf4d1dC1eCe984",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x3Deb42367eEf8b2D895689aa313f0eA9Eb679198",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x6B1d09A97b813D53e9D4b7523DA36604C0B52242",
                "0x9e551B6694Bbd295D7d6e6A2540C7d41Ce70a3b9",
                "0xbFB8911b72CfB138C7Ce517c57d9C691535DC517"
              ],
              "threshold": 2
            },
            {
              "address": "0xA07B4d80F7d6ce8358A043dfaB23615969399174",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x6B1d09A97b813D53e9D4b7523DA36604C0B52242",
                "0x9e551B6694Bbd295D7d6e6A2540C7d41Ce70a3b9",
                "0xbFB8911b72CfB138C7Ce517c57d9C691535DC517"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "celo": {
          "address": "0x0B8f51272a04e0AcBD901C956A770A046890816e",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x210130063fFeCbBb783037882cFF907c78CB2e77",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x46eCbc794574727AbB8f97F01DAcD9db6135F47A",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x622e43BAF06ad808cA8399360D9a2d9A1a12688b",
                "0x63478422679303c3e4fC611B771fa4a707eF7f4a",
                "0xF2c1e3888eb618F1f1a071eF3618f134715a9a49"
              ],
              "threshold": 2
            },
            {
              "address": "0xFB0448b8649cB50bF5163c313afd4eb86eA981D9",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x46eCbc794574727AbB8f97F01DAcD9db6135F47A",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x622e43BAF06ad808cA8399360D9a2d9A1a12688b",
                "0x63478422679303c3e4fC611B771fa4a707eF7f4a",
                "0xF2c1e3888eb618F1f1a071eF3618f134715a9a49"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "ethereum": {
          "address": "0x9e2DF2BB8c77bBb0cF77a37f8a03fa1788C202Ab",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0xD76BFda42e975EEC377F87c4513374eEc8F1b8A8",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x03c842db86A6A3E524D4a6615390c1Ea8E2b9541",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x94438A7De38D4548ae54dF5c6010c4Ebc5239eAE",
                "0xb2f5a6A6e6046E2EDE213617e989329666a6C4bc",
                "0xCE327111035dd38698c92c3778884dBbB0ca8103"
              ],
              "threshold": 3
            },
            {
              "address": "0xb9f12Cc11b3e36d7914E60C0C290f93c01a1AC85",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x03c842db86A6A3E524D4a6615390c1Ea8E2b9541",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x94438A7De38D4548ae54dF5c6010c4Ebc5239eAE",
                "0xb2f5a6A6e6046E2EDE213617e989329666a6C4bc",
                "0xCE327111035dd38698c92c3778884dBbB0ca8103"
              ],
              "threshold": 3
            }
          ],
          "threshold": 1
        },
        "polygonzkevm": {
          "address": "0x712630582473a57075F475c81a5510D55E836F6d",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0xBef1Bf4EE74816a690cb71DAC231B3172FBc4bDf",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x57231619fEA13D85270Ca6943298046C75a6dD01",
                "0x865818Fe1dB986036d5FD0466dCd462562436d1A",
                "0x86f2A44592bB98dA766E880cfd70D3bbb295e61a"
              ],
              "threshold": 2
            },
            {
              "address": "0xF9F4D14D7869964fA5B52077A7C91FeB09B2C038",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x57231619fEA13D85270Ca6943298046C75a6dD01",
                "0x865818Fe1dB986036d5FD0466dCd462562436d1A",
                "0x86f2A44592bB98dA766E880cfd70D3bbb295e61a"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "base": {
          "address": "0xCBc40857e42FA4896a3b1f19ce3b6e383D4bF693",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x90f0F3C950D796F31e7261e0e5386A00Df4C272b",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x41188CB5A5493a961c467BA38a3F8b1F1D35eE63",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x64889519aD3ffc8F3aE436fCD63efc6b853fd63f",
                "0xB9453D675E0Fa3c178A17B4cE1AD5B1A279B3aF9",
                "0xCff391b4E516452d424db66beb9052b041a9eD79"
              ],
              "threshold": 2
            },
            {
              "address": "0xb2965fe19d3ae1821F903DA8A452C9e1F149B5cC",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x41188CB5A5493a961c467BA38a3F8b1F1D35eE63",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x64889519aD3ffc8F3aE436fCD63efc6b853fd63f",
                "0xB9453D675E0Fa3c178A17B4cE1AD5B1A279B3aF9",
                "0xCff391b4E516452d424db66beb9052b041a9eD79"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "neutron": {
          "address": "0x32E905cF3480e7674087bF69b475e3E54166a1d3",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x4d7C0FAaB90FdAC50335e1De37BE8Fd3763CAc01",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x42B6De2edBaA62C2eA2309Ad85D20B3E37d38Acf",
                "0x42FA752DEFe92459370A052B6387a87f7DE9b80C",
                "0x47AA126E05933b95c5eB90b26e6b668D84F4b25A",
                "0x54B2cca5091B098A1A993dEC03c4d1EE9AF65999",
                "0xA9b8c1f4998f781f958C63cfCD1708d02f004fF0",
                "0xb65438a014FB05fbADCFE35bC6e25D372b6Ba460",
                "0xC79503a3e3011535a9c60f6d21F76F59823A38bD"
              ],
              "threshold": 4
            },
            {
              "address": "0x718E0bc33c50DB90ef27A97B37E4580c9Fa1bbF3",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x42B6De2edBaA62C2eA2309Ad85D20B3E37d38Acf",
                "0x42FA752DEFe92459370A052B6387a87f7DE9b80C",
                "0x47AA126E05933b95c5eB90b26e6b668D84F4b25A",
                "0x54B2cca5091B098A1A993dEC03c4d1EE9AF65999",
                "0xA9b8c1f4998f781f958C63cfCD1708d02f004fF0",
                "0xb65438a014FB05fbADCFE35bC6e25D372b6Ba460",
                "0xC79503a3e3011535a9c60f6d21F76F59823A38bD"
              ],
              "threshold": 4
            }
          ],
          "threshold": 1
        },
        "moonbeam": {
          "address": "0x9bD9864FD3E17008764dA2dC9cb96B77FD44eDd8",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x3146517E6a45227604745338FDBEa2cF2e64e5c4",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x2225e2F4e9221049456Da93B71D2De41f3B6b2A8",
                "0x645428D198d2E76cbD9c1647F5c80740BB750B97",
                "0xaED886392df07897743D8E272d438F00c4c9a2Ae",
                "0xCF0bb43255849cB3709a96EE166E5c3ce4AdC7F9"
              ],
              "threshold": 2
            },
            {
              "address": "0xD7039fe2703CD5043708d25912f763D93ca850A6",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x2225e2F4e9221049456Da93B71D2De41f3B6b2A8",
                "0x645428D198d2E76cbD9c1647F5c80740BB750B97",
                "0xaED886392df07897743D8E272d438F00c4c9a2Ae",
                "0xCF0bb43255849cB3709a96EE166E5c3ce4AdC7F9"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "gnosis": {
          "address": "0x157623a1B5621a5aFFdCE20d73De71a24dfd8933",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x5B005ebad07C8635B5FdA58095291FDf39fD8181",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x19fB7E04A1Be6b39b6966a0b0C60B929A93eD672",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0xd4dF66a859585678f2Ea8357161D896BE19cC1Ca",
                "0xDB96116d13a2FaDde9742d7cc88474a5ED39A03A"
              ],
              "threshold": 2
            },
            {
              "address": "0xcD538D071b62F1F8fa99586a01D0403Bca67E1Fe",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x19fB7E04A1Be6b39b6966a0b0C60B929A93eD672",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0xd4dF66a859585678f2Ea8357161D896BE19cC1Ca",
                "0xDB96116d13a2FaDde9742d7cc88474a5ED39A03A"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        },
        "polygon": {
          "address": "0x40236cedbC4FFbbE7601922f20bB59aE9Fff3dD2",
          "type": "staticAggregationIsm",
          "modules": [
            {
              "address": "0x47Da9f405b0ff0E410C787c7210C6Fa9d2ecA3b1",
              "type": "merkleRootMultisigIsm",
              "validators": [
                "0x008F24Cbb1cc30ad0f19f2516CA75730e37EfB5f",
                "0x12ECb319C7f4e8ac5eB5226662aeB8528c5cefac",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x722Aa4D45387009684582BCA8281440D16b8b40F"
              ],
              "threshold": 2
            },
            {
              "address": "0x62DA20A9be5538626df0Cb456d1E1311Cb5Ef29d",
              "type": "messageIdMultisigIsm",
              "validators": [
                "0x008F24Cbb1cc30ad0f19f2516CA75730e37EfB5f",
                "0x12ECb319C7f4e8ac5eB5226662aeB8528c5cefac",
                "0x5450447aeE7B544c462C9352bEF7cAD049B0C2Dc",
                "0x722Aa4D45387009684582BCA8281440D16b8b40F"
              ],
              "threshold": 2
            }
          ],
          "threshold": 1
        }
      }
    }
  ],
  "threshold": 2
}
```

File: typescript/sdk/src/hook/read.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulbalaji) [reply typescript/sdk/src/hook/read.ts:144]: this is a great find, thanks

File: typescript/sdk/src/utils/concurrentMap.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Map an async function over a list xs with a given concurrency level
+ * Forked from https://github.com/celo-org/developer-tooling/blob/0c61e7e02c741fe10ecd1d733a33692d324cdc82/packages/sdk/base/src/async.ts#L128
+ *
+ * @param concurrency number of `mapFn` concurrent executions
+ * @param xs list of value
+ * @param mapFn mapping function
+ */
+export async function concurrentMap<A, B>(
```
Comment (jmrossy) [inline typescript/sdk/src/utils/concurrentMap.ts:9]: This should be in the utils/src/async.ts file

File: typescript/sdk/src/ism/read.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+import {
+  Address,
+  WithAddress,
+  assert,
+  ethersBigNumberSerializer,
+  rootLogger,
+} from '@hyperlane-xyz/utils';
+
+import { Chains } from '../consts/chains.js';
+import { concurrentMap } from '../index.js';
```
Comment (jmrossy) [inline typescript/sdk/src/ism/read.ts:21]: Nit: prefer direct imports to these than go through the index file. These roundabout imports can reduce the efficacy of tree shaking and other optimizations

File: typescript/sdk/src/ism/read.ts [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+  ): Promise<WithAddress<PausableIsmConfig | TestIsmConfig | OpStackIsmConfig>>;
+}
+
+export class EvmIsmReader implements IsmReader {
+  protected readonly provider: providers.Provider;
+  protected readonly logger = rootLogger.child({ module: 'EvmIsmReader' });
+
+  constructor(
+    protected readonly multiProvider: MultiProvider,
+    chain: Chains,
```
Comment (jmrossy) [inline typescript/sdk/src/ism/read.ts:56]: I think you mean `chain: ChainName` here.

**Review by jmrossy — APPROVED** [review]
A few minor things but overall lgtm

File: typescript/sdk/src/ism/read.ts (hunk 5 above)
Comment (paulbalaji) [reply typescript/sdk/src/ism/read.ts:21]: woops not intentional, nice catch

**Pull request merged**